File: 3.x.x/crates-utopia/messaging/src/adapter/sms/telesign.rs
//! PHP `Utopia\Messaging\Adapter\SMS\Telesign`.

use serde_json::json;

use super::TYPE;
use crate::adapter::{expect_sms, Adapter, AdapterBase, SendResult};
use crate::error::MessagingError;
use crate::message::{Message, MessageKind};
use crate::messages::SMS;
use crate::php::basic_auth;
use crate::response::{Response, ResponseData};

/// PHP `Adapter\SMS\Telesign`.
#[derive(Debug)]
pub struct Telesign {
    base: AdapterBase,
    customer_id: String,
    api_key: String,
}

impl Telesign {
    /// PHP `__construct($customerId, $apiKey)`.
    #[must_use]
    pub fn new(customer_id: impl Into<String>, api_key: impl Into<String>) -> Self {
        Self {
            base: AdapterBase::default(),
            customer_id: customer_id.into(),
            api_key: api_key.into(),
        }
    }

    fn format_numbers(numbers: &[String]) -> String {
        numbers
            .iter()
            .map(|n| format!("{n}:{}", uniqid()))
            .collect::<Vec<_>>()
            .join(",")
    }

    fn process_sms(&self, message: &SMS) -> ResponseData {
        let to = Self::format_numbers(message.get_to());
        let result = self.request_default(
            "POST",
            "https://rest-ww.telesign.com/v1/verify/bulk_sms",
            &[
                "Content-Type: application/x-www-form-urlencoded".into(),
                format!(
                    "Authorization: Basic {}",
                    basic_auth(&self.customer_id, &self.api_key)
                ),
            ],
            Some(json!({
                "template": message.get_content(),
                "recipients": to,
            })),
        );
        let mut response = Response::new(TYPE);
        if result.status_code == 200 {
            response.set_delivered_to(message.get_to().len() as i64);
            for dest in message.get_to() {
                response.add_result(dest, "");
            }
        } else {
            let desc = result
                .response
                .get("errors")
                .and_then(|e| e.as_array())
                .and_then(|e| e.first())
                .and_then(|e| e.get("description"))
                .and_then(|v| v.as_str());
            for dest in message.get_to() {
                response.add_result(dest, desc.unwrap_or("Unknown error"));
            }
        }
        response.to_array()
    }
}

fn uniqid() -> String {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    format!("{:x}{:x}", rng.gen::<u32>(), rng.gen::<u16>())
}

impl Adapter for Telesign {
    fn get_name(&self) -> &'static str {
        "Telesign"
    }
    fn get_type(&self) -> &'static str {
        TYPE
    }
    fn get_message_type(&self) -> MessageKind {
        MessageKind::SMS
    }
    fn get_max_messages_per_request(&self) -> usize {
        1000
    }
    fn base(&self) -> &AdapterBase {
        &self.base
    }
    fn process(&self, message: &dyn Message) -> Result<SendResult, MessagingError> {
        Ok(SendResult::Response(self.process_sms(expect_sms(message)?)))
    }
}

File: packages/messaging/src/Utopia/Messaging/Adapter/SMS/Telesign.php
<?php

namespace Utopia\Messaging\Adapter\SMS;

use Utopia\Messaging\Adapter\SMS as SMSAdapter;
use Utopia\Messaging\Messages\SMS as SMSMessage;
use Utopia\Messaging\Response;

// Reference Material
// https://developer.telesign.com/enterprise/reference/sendbulksms

class Telesign extends SMSAdapter
{
    protected const NAME = 'Telesign';

    /**
     * @param  string  $customerId Telesign customer ID
     * @param  string  $apiKey Telesign API key
     */
    public function __construct(
        private readonly string $customerId,
        private readonly string $apiKey,
    ) {
        parent::__construct();
    }

    public function getName(): string
    {
        return static::NAME;
    }

    public function getMaxMessagesPerRequest(): int
    {
        return 1000;
    }

    /**
     * {@inheritdoc}
     *
     * @throws \Exception
     */
    protected function process(SMSMessage $message): array
    {
        $to = $this->formatNumbers(array_map(
            fn(string $to): string => $to,
            $message->getTo(),
        ));

        $response = new Response($this->getType());

        $result = $this->request(
            method: 'POST',
            url: 'https://rest-ww.telesign.com/v1/verify/bulk_sms',
            headers: [
                'Content-Type: application/x-www-form-urlencoded',
                'Authorization: Basic ' . base64_encode("{$this->customerId}:{$this->apiKey}"),
            ],
            body: [
                'template' => $message->getContent(),
                'recipients' => $to,
            ],
        );

        if ($result['statusCode'] === 200) {
            $response->setDeliveredTo(\count($message->getTo()));
            foreach ($message->getTo() as $to) {
                $response->addResult($to);
            }
        } else {
            foreach ($message->getTo() as $to) {
                if (!\is_null($result['response']['errors'][0]['description'] ?? null)) {
                    $response->addResult($to, $result['response']['errors'][0]['description']);
                } else {
                    $response->addResult($to, 'Unknown error');
                }
            }
        }

        return $response->toArray();
    }

    /**
     * @param  array<string>  $numbers
     */
    private function formatNumbers(array $numbers): string
    {
        $formatted = array_map(
            fn(string $number): string => $number . ':' . uniqid(),
            $numbers,
        );

        return implode(',', $formatted);
    }
}
